src/pymor/algorithms/phdmd.py

```diff
@@ -0,0 +1,468 @@
+    atol
+        Finish when the absolute error measure is below this threshold.
+    rtol
+        Finish when the relative update error measure is below this threshold.
+    weighted_rtol
+        In the weighted pH DMD initialization problem, truncate when the relative value is
+        below this threshold.
+    skew_procrustes_rtol
+        In the skew-symmetric Procrustes problem, truncate when the relative value is
+        below this threshold.
+
+    Returns
+    -------
+    J
+        The `NumPy array` containing :math:`J`.
+    R
+        The `NumPy array` containing :math:`R`.
+    data
+        Dict containing the following fields:
+
+            :update_norms:          |NumPy array| of the norms of the update vectors after
+                                    each iteration.
+            :abs_errs:              |NumPy array| of the absolute errors after each iteration.
+            :rel_errs:              |NumPy array| of the relative  errors after each iteration.
+            :iterations:            Number of total pH DMD iterations.
+            :skew_procrustes_data:  `list` of data generated by the skew-symmetric
+                                    Procrustes subproblems.
+            :weighted_init_data:    If `initial_J` and `initial_R` are both `None`, dict of data
+                                    generated by the weighted pH DMD initialization subproblem.
+            :skew_init_data:        If only `initial_J` is `None`, dict of data generated by the
+                                    skew-symmetric Procrustes subproblem.
+            :spsd_init_data:        If only `initial_R` is `None`, dict of data generated by the psd
+                                    symmetric Procrustes subproblem.
+    """
+    assert isinstance(X, VectorArray)
+    assert isinstance(Y, VectorArray)
+    assert isinstance(U, VectorArray)
+    assert len(X) == len(Y)
+    assert len(X) % len(U) == len(Y) % len(U) == 0
+    assert 0. < initial_alpha < 1.
+
+    logger = getLogger('pymor.algorithms.phdmd.phdmd')
+
+    if Xdot is None:
+        assert dt is not None
+        Xdot = (X[1:] - X[:-1]) * (1. / dt)
+    else:
+        assert isinstance(Xdot, VectorArray)
+        assert len(X) == len(Xdot)
+        assert len(X) == len(U)
+        Xdot = .5 * (Xdot[1:] + Xdot[:-1])
+
+    X = .5 * (X[1:] + X[:-1])
+    Y = .5 * (Y[1:] + Y[:-1])
+    U = .5 * (U[1:] + U[:-1])
+
+    if H is None:
+        logger.warn('No H matrix provided. Did you intend this?')
+        H = np.eye(X.dim)
+    else:
+        if isinstance(H, Operator):
+            assert H.source == H.range == X.space
+            H = to_matrix(H)
+        elif isinstance(H, VectorArray):
+            assert len(H) == H.dim == X.dim
+            H = H.to_numpy()
+
+    assert isinstance(H, np.ndarray)
+    assert len(H.shape) == 2
+
+    Xdot = Xdot.to_numpy().T
+    X = X.to_numpy().T
+    Y = Y.to_numpy().T
+    U = U.to_numpy().T
```

Comment:
This kind of misses the whole point of using `VectorArrays`... Can the algorithm be reformulated to only use the `VectorArray` interface? If not, how would we justify adding this to pyMOR?

Reply:
I rewrote the algorithm using `VectorArrays` and `BlockOperators`. I'm unsure however if the ordering of the `BlockVectorSpace` subspaces always stays the same, so I'd appreciate feedback in respect to their utilization.

Reply:
Ordering should be fixed (if it's not, that sounds like a bug).
I still found in a few places that `to_numpy` or `to_matrix` is used (https://github.com/pymor/pymor/pull/2027#discussion_r1272706606). If it is not possible to avoid this, the first step in the algorithm could be dimensionality reduction using POD (or DMD).